### DISCUSSION TITLE
Document available depTypes

### DISCUSSION THREAD

**Discussion by SebastianGoeb** [renovatebot/renovate#15176] [Suggest an Idea]
Different managers support different depTypes in packageRules, but it seems like the available strings aren't documented anywhere. 

The [NPM manager docs](https://docs.renovatebot.com/modules/manager/npm/) do mention depTypes, but the [maven docs](https://docs.renovatebot.com/modules/manager/maven/), for example, don't. I tried looking in the source code, and after some searching I stumbled upon [npm/extract/index.ts](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/npm/extract/index.ts#L161) and [maven/extract.ts](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/maven/extract.ts#L72). So it seems the implementation isn't super consistent wrt. file structure, and whether or not a module even declares the complete list of depTypes anywhere.

Is it possible to maybe make this more consistent and document the available options? At the moment I find it hard to use the packageRules/matchDepTypes feature.

**Comment by rarkins** [comment]
I thought we had an issue open for this, but I'm not able to find it.

I think it's a good idea, but with a couple of challenges:
1. We ideally want to do this using code, not pure docs. Docs can easily get out of sync. So e.g. we want to have each manager declare its possible `depType` values and then we enforce it with tests or linting somehow
2. For some managers, `depType` values can be arbitrary and aren't a fixed set, because they are determined from user settings within the package file. In this case such docs would not help your use case because at best they can just say "any depType"

**Reply by viceice** [reply]
i think those should be on manager docs, so every manager can documented it separately.

we can also  add a runtime check/ log it there is a mismatch

**Reply by SebastianGoeb** [reply]
That would be very cool. Do you have any code-generated docs already? Cause I couldn't find anything at a glance. The NPM manager has the depTypes documented in a standard README.md

In my current project we use asciidoc instead of markdown, because that lets us include snippets of text from the code into the docs. That way the docs are still handwritten, but at least they're right next to the implementation and less likely to become outdated.

**Reply by rarkins** [reply]
Much of the manager docs are already automatically generated from the manager exports as part of our build process

**Reply by dsegurag** [reply]
hey! is there any way to check them via code or any other method then? thanks in advance.

**Reply by jamietanna** [reply]
@dsegurag are you looking at _all possible_ `depType` or `depTypes` that could be known to Renovate? Or what's available to your project?

If the latter, checking the Renovate debug logs will surface a `packageFiles with updates` message, which contains the internal state Renovate has for the known package updates, and that indicates the `depType`/`depTypes` that are known for the dependencies

**Comment by woile** [comment]
This would be really useful, I cannot find a single mention of python `depTypes`, is it `devDependencies, dev-dependencies, optional-dependnecies, optionalDependencies, tool.uv.dev-dependencies`?

**Reply by jamietanna** [reply]
The best way (at least right now, but my usual means for checking this) is to check your debug logs for `packageFiles with updates`, and see what `depType` or `depTypes` are produced by the various dependencies

For instance, a snippet from Renovate's own logs:

```
    "pep621": [
      {
        "extractedConstraints": {
          "python": ">=3.11"
        },
        "deps": [
          {
            "packageName": "python",
            "depType": "requires-python",
            "currentValue": ">=3.11",
            "commitMessageTopic": "Python",
            "datasource": "python-version",
            "versioning": "pep440",
            "depName": "python",
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/python/cpython",
            "registryUrl": "https://www.python.org/api/v2/downloads/release",
            "homepage": "https://python.org",
            "currentVersion": "3.14.3",
            "currentVersionTimestamp": "2026-02-03T18:27:43.000Z",
            "currentVersionAgeInDays": 47
          },
          {
            "packageName": "mkdocs-material",
            "depName": "mkdocs-material",
            "datasource": "pypi",
            "depType": "project.dependencies",
            "currentValue": "==9.7.6",
            "currentVersion": "9.7.6",
            "lockedVersion": "9.7.6",
            "updates": [],
            "versioning": "pep440",
            "warnings": [],
            "sourceUrl": "https://github.com/squidfunk/mkdocs-material",
            "registryUrl": "https://pypi.org/pypi",
            "changelogUrl": "https://squidfunk.github.io/mkdocs-material/changelog/",
            "mostRecentTimestamp": "2026-03-19T15:41:55.000Z",
            "isAbandoned": false,
            "currentVersionTimestamp": "2026-03-19T15:41:55.000Z",
            "currentVersionAgeInDays": 3,
            "fixedVersion": "9.7.6"
          },
          {
            "packageName": "mkdocs-awesome-pages-plugin",
            "depName": "mkdocs-awesome-pages-plugin",
            "datasource": "pypi",
            "depType": "project.dependencies",
            "currentValue": "==2.10.1",
            "currentVersion": "2.10.1",
            "lockedVersion": "2.10.1",
            "updates": [],
            "versioning": "pep440",
            "warnings": [],
            "sourceUrl": "https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin",
            "registryUrl": "https://pypi.org/pypi",
            "mostRecentTimestamp": "2024-12-22T21:13:46.000Z",
            "isAbandoned": true,
            "currentVersionTimestamp": "2024-12-22T21:13:46.000Z",
            "currentVersionAgeInDays": 455,
            "fixedVersion": "2.10.1"
          }
        ],
        "lockFiles": [
          "pdm.lock"
        ],
        "packageFile": "pyproject.toml"
      }
    ],
```

**Reply by woile** [reply]
Thanks, I used this command in the end to find them:

```sh
LOG_LEVEL=debug npx --yes renovate --dry-run=full --platform=local
```

**Comment by jamietanna** [comment]
I'd recently implemented this - each Manager's documentation will have a `Dependency types` header

i.e. [Ansible](https://docs.renovatebot.com/modules/manager/ansible/#dependency-types) (nothing documented) vs [setup-cfg](https://docs.renovatebot.com/modules/manager/setup-cfg/) (which does have)
